### ISSUE TITLE
support launching Intel MPI jobs on Intel OPA fabric

### ISSUE THREAD

**Issue by garlick** [flux-framework/flux-core#630]
For Flux testing on TOSS3/RHEL7 on the CTS-1 hardware, Intel MPI support is desirable as presently only Intel MPI and Open MPI support the CTS-1 interconnect.

Looks like it works (setting either I_MPI_PMI_LIBRARY or LD_PRELOAD works).  This is on a 2 node flux session launched with slurm (mpi procs are one per node):
```
[garlick@upgrade7:mpi]$ module load intel-mpi
[garlick@upgrade7:mpi]$  cc -I/opt/intel-mpi/5.1/include -o hello hello.c -L/opt/intel-mpi/5.1/lib -lmpi
[garlick@upgrade7:mpi]$ PMI_TRACE=0xff  I_MPI_PMI_LIBRARY=/g/g0/garlick/flux-core/src/lib/libpmi/.libs/libpmi.so flux wreckrun --ntasks 2 ../../t/mpi/hello
0: completed first barrier
0: completed MPI_Init in 0.181s.  There are 2 tasks
0: completed MPI_Finalize
```

PMI trace output, sorted by timestamp
```
[1459891306.455442] libpmi.debug[0]: PMI_Init = SUCCESS
[1459891306.455478] libpmi.debug[0]: PMI_KVS_Get_my_name = SUCCESS
[1459891306.455715] libpmi.debug[0]: PMI_KVS_Put (lwj.2.pmi.P0-hostname = "upgrade7") = SUCCESS
[1459891306.455733] libpmi.debug[0]: PMI_KVS_Commit = SUCCESS
[1459891306.499622] libpmi.debug[1]: PMI_Init = SUCCESS
[1459891306.499665] libpmi.debug[1]: PMI_KVS_Get_my_name = SUCCESS
[1459891306.499855] libpmi.debug[1]: PMI_KVS_Put (lwj.2.pmi.P1-hostname = "upgrade8") = SUCCESS
[1459891306.499870] libpmi.debug[1]: PMI_KVS_Commit = SUCCESS
[1459891306.500448] libpmi.debug[0]: PMI_Barrier (lwj.2.pmi:0, 2) = SUCCESS
[1459891306.500668] libpmi.debug[1]: PMI_Barrier (lwj.2.pmi:0, 2) = SUCCESS
[1459891306.500689] libpmi.debug[0]: PMI_KVS_Get (lwj.2.pmi.P1-hostname = "upgrade8") = SUCCESS
[1459891306.500939] libpmi.debug[0]: PMI_KVS_Put (lwj.2.pmi.r2h0 = "thnum#2$h0#upgrade7$r0#0$h1#upgrade8$r1#1$nbc#1$") = SUCCESS
[1459891306.500962] libpmi.debug[0]: PMI_KVS_Commit = SUCCESS
[1459891306.501410] libpmi.debug[0]: PMI_Barrier (lwj.2.pmi:1, 2) = SUCCESS
[1459891306.501426] libpmi.debug[0]: PMI_Get_rank = SUCCESS
[1459891306.501439] libpmi.debug[0]: PMI_Get_size = SUCCESS
[1459891306.501447] libpmi.debug[0]: PMI_Get_appnum = SUCCESS
[1459891306.501455] libpmi.debug[0]: PMI_KVS_Get_name_length_max = SUCCESS
[1459891306.501467] libpmi.debug[0]: PMI_KVS_Get_my_name = SUCCESS
[1459891306.501490] libpmi.debug[0]: PMI_KVS_Get_name_length_max = SUCCESS
[1459891306.501507] libpmi.debug[0]: PMI_KVS_Get_my_name = SUCCESS
[1459891306.501598] libpmi.debug[1]: PMI_Barrier (lwj.2.pmi:1, 2) = SUCCESS
[1459891306.502479] libpmi.debug[0]: PMI_KVS_Get_value_length_max = SUCCESS
[1459891306.502525] libpmi.debug[0]: PMI_KVS_Get_key_length_max = SUCCESS
[1459891306.502539] libpmi.debug[0]: PMI_KVS_Get_value_length_max = SUCCESS
[1459891306.502594] libpmi.debug[1]: PMI_KVS_Get (lwj.2.pmi.r2h0 = "thnum#2$h0#upgrade7$r0#0$h1#upgrade8$r1#1$nbc#1$") = SUCCESS
[1459891306.502623] libpmi.debug[1]: PMI_Get_rank = SUCCESS
[1459891306.502642] libpmi.debug[1]: PMI_Get_size = SUCCESS
[1459891306.502657] libpmi.debug[1]: PMI_Get_appnum = SUCCESS
[1459891306.502673] libpmi.debug[1]: PMI_KVS_Get_name_length_max = SUCCESS
[1459891306.502689] libpmi.debug[1]: PMI_KVS_Get_my_name = SUCCESS
[1459891306.502713] libpmi.debug[1]: PMI_KVS_Get_name_length_max = SUCCESS
[1459891306.502727] libpmi.debug[1]: PMI_KVS_Get_my_name = SUCCESS
[1459891306.503543] libpmi.debug[1]: PMI_KVS_Get_value_length_max = SUCCESS
[1459891306.503587] libpmi.debug[1]: PMI_KVS_Get_key_length_max = SUCCESS
[1459891306.503599] libpmi.debug[1]: PMI_KVS_Get_value_length_max = SUCCESS
[1459891306.503867] libpmi.debug[0]: PMI_Barrier (lwj.2.pmi:2, 2) = SUCCESS
[1459891306.504064] libpmi.debug[1]: PMI_Barrier (lwj.2.pmi:2, 2) = SUCCESS
[1459891306.611885] libpmi.debug[1]: PMI_KVS_Put (lwj.2.pmi.P1-businesscard-0 = "rdma_port0#30104$rdma_host0#0A00000E00000FE4FE80000000000000F45214030057AD2100000004$arch_code#0$fabrics_list#dapl$") = SUCCESS
[1459891306.611902] libpmi.debug[1]: PMI_KVS_Commit = SUCCESS
[1459891306.612932] libpmi.debug[0]: PMI_KVS_Put (lwj.2.pmi.P0-businesscard-0 = "rdma_port0#31860$rdma_host0#0A00000700001006FE80000000000000F45214030057944100000004$arch_code#0$fabrics_list#dapl$") = SUCCESS
[1459891306.612948] libpmi.debug[0]: PMI_KVS_Commit = SUCCESS
[1459891306.613408] libpmi.debug[0]: PMI_Barrier (lwj.2.pmi:3, 2) = SUCCESS
[1459891306.613590] libpmi.debug[0]: PMI_KVS_Get_value_length_max = SUCCESS
[1459891306.613702] libpmi.debug[1]: PMI_Barrier (lwj.2.pmi:3, 2) = SUCCESS
[1459891306.613803] libpmi.debug[0]: PMI_KVS_Get (lwj.2.pmi.P1-businesscard-0 = "rdma_port0#30104$rdma_host0#0A00000E00000FE4FE80000000000000F45214030057AD2100000004$arch_code#0$fabrics_list#dapl$") = SUCCESS
[1459891306.613866] libpmi.debug[1]: PMI_KVS_Get_value_length_max = SUCCESS
[1459891306.617467] libpmi.debug[0]: PMI_Barrier (lwj.2.pmi:4, 2) = SUCCESS
[1459891306.617660] libpmi.debug[1]: PMI_Barrier (lwj.2.pmi:4, 2) = SUCCESS
[1459891306.620644] libpmi.debug[0]: PMI_Finalize = SUCCESS
[1459891306.620778] libpmi.debug[1]: PMI_Finalize = SUCCESS
```

**Comment by garlick** [comment]
First try on opal not successful:
```
bash-4.2$ export I_MPI_PMI_LIBRARY=/g/g0/garlick/flux-core/src/lib/libpmi/.libs/libpmi.so
bash-4.2$ flux wreckrun --ntasks 2 t/mpi/hello
opal37:SCM:12df9:aab0a300: 31 us(31 us):  open_hca: device mlx4_0 not found
opal37:SCM:12df9:aab0a300: 40 us(40 us):  open_hca: device mlx4_0 not found
opal38:SCM:9a3f:aab0a300: 40 us(40 us):  open_hca: device mlx4_0 not found
opal37:CMA:12df9:aab0a300: 40 us(40 us):  open_hca: getaddr_netdev ERROR:No such device. Is ib0 configured?
opal37:CMA:12df9:aab0a300: 55 us(55 us):  open_hca: getaddr_netdev ERROR:No such device. Is ib1 configured?
[snip]
```

**Comment by garlick** [comment]
On opal:  retried under srun, and got the same errors.  Trent provided this:
```
Date: Thu, 7 Apr 2016 13:26:19 -0700
From: Trent D'Hooge <tdhooge@llnl.gov>
To: "Garlick, Jim E." <garlick1@llnl.gov>
Subject: intel tmi variables

module load intel-mpi
export I_MPI_FABRICS_LIST=tmi
export TMI_CONFIG=/opt/intel-mpi/5.1/etc/tmi.conf
```
with these set, srun began working with Intel MPI.  Here are some Slurm timing results for comparison with results in #625.  8192 tasks segfaulted.
```
0: completed MPI_Init in 1.406s.  There are 128 tasks
0: completed first barrier
0: completed MPI_Finalize

real	0m4.629s
user	0m0.287s
sys	0m0.358s
srun: job 1004672 queued and waiting for resources
srun: job 1004672 has been allocated resources
0: completed MPI_Init in 1.921s.  There are 256 tasks
0: completed first barrier
0: completed MPI_Finalize

real	0m10.924s
user	0m0.307s
sys	0m0.435s
srun: job 1004673 queued and waiting for resources
srun: job 1004673 has been allocated resources
0: completed MPI_Init in 2.066s.  There are 512 tasks
0: completed first barrier
0: completed MPI_Finalize

real	0m16.357s
user	0m0.407s
sys	0m0.626s
srun: job 1004674 queued and waiting for resources
srun: job 1004674 has been allocated resources
0: completed MPI_Init in 3.887s.  There are 1024 tasks
0: completed first barrier
0: completed MPI_Finalize

real	0m19.374s
user	0m0.535s
sys	0m1.122s
srun: job 1004675 queued and waiting for resources
srun: job 1004675 has been allocated resources
0: completed MPI_Init in 7.013s.  There are 2048 tasks
0: completed first barrier
0: completed MPI_Finalize

real	0m34.696s
user	0m0.929s
sys	0m2.251s
srun: job 1004676 queued and waiting for resources
srun: job 1004676 has been allocated resources
0: completed MPI_Init in 14.906s.  There are 4096 tasks
0: completed first barrier
0: completed MPI_Finalize

real	1m7.766s
user	0m1.143s
sys	0m2.751s
srun: job 1004677 queued and waiting for resources
srun: job 1004677 has been allocated resources
srun: error: opal71: tasks 2177-2178,2180,2184,2191: Segmentation fault (core dumped)
srun: error: opal77: tasks 2594-2596,2600,2607: Segmentation fault (core dumped)
```

**Comment by garlick** [comment]
When I run with these runes under Flux, I get a hang.  strace and pmi tracing seem to indicate that we get through the PMI part of MPI_Init() but must get stuck somewhere after that.
```
rt_sigaction(SIGBUS, {0x2aaaafeb70d0, [BUS], SA_RESTORER|SA_RESTART, 0x2aaaabe99670}, {0x2aaaafe84a20, [BUS], SA_RESTORER|SA_RESTART, 0x2aaaabe99670}, 8) = 0
process_vm_readv(83818, [], 0, [], 0, 0) = 0
fcntl(3, F_GETFL)                       = 0x2 (flags O_RDWR)
fcntl(3, F_SETFL, O_RDWR)               = 0
write(3, "\0\0\0\264", 4)               = 4
write(3, "\0\10cmb.log\0\234{ \"facility\": \"libpmi"..., 180) = 180
fcntl(3, F_GETFL)                       = 0x2 (flags O_RDWR)
fcntl(3, F_SETFL, O_RDWR)               = 0
write(3, "\0\0\0A", 4)                  = 4
write(3, "\0\10kvs.get\0){ \"lwj.2.pmi.P0-busin"..., 65) = 65
fcntl(3, F_GETFL)                       = 0x2 (flags O_RDWR)
fcntl(3, F_SETFL, O_RDWR)               = 0
read(3, "\0\0\0@", 4)                   = 4
read(3, "\10kvs.get\0){ \"lwj.2.pmi.P0-busine"..., 64) = 64
```

**Comment by garlick** [comment]
Reopening, since we still need to resolve the above hang on the CTS-1 hardware.

**Comment by garlick** [comment]
Still hanging.  Here's the makefile I used to build/launch MPI hello
```makefile
top_intel_dir = /opt/intel-mpi/5.1
CC=$(top_intel_dir)/bin/mpicc
CFLAGS=-I$(top_intel_dir)/include
LDFLAGS=-L$(top_intel_dir)/lib

all: hello.intel

hello.intel: hello.c
	$(CC) $(CFLAGS) -o $@ $< $(LDFLAGS)

run: hello.intel
	export LD_LIBRARY_PATH=$(top_intel_dir)/lib:$LD_LIBRARY_PATH; \
	export I_MPI_FABRICS_LIST=tmi; \
	export TMI_CONFIG=$(top_intel_dir)/etc/tmi.conf; \
	PMI_TRACE=0xff flux wreckrun -l --ntasks 2 strace hello.intel
```
rank 0 strace (tail)
```
0: write(3, "\0\0\0C", 4)                  = 4
0: write(3, "\0\10kvs.put\0+{ \"lwj.8.pmi.TMI_PROV"..., 67) = 67
0: read(3, "\0\0\0\26", 4)                 = 4
0: read(3, "\10kvs.put\0\f\216\1\2\1\0\0\0\0\0\0\0\1", 22) = 22
0: write(3, "\0\0\0\315", 4)               = 4
0: write(3, "\0\10cmb.log\0\265{ \"facility\": \"libpmi"..., 205) = 205
0: write(3, "\0\0\0\246", 4)               = 4
0: write(3, "\0\10cmb.log\0\216{ \"facility\": \"libpmi"..., 166) = 166
0: write(3, "\0\0\0Z", 4)                  = 4
0: write(3, "\0\vkvs.commit\0?{ \"ops\": null, \".a"..., 90) = 90
0: read(3, "\0\0\0\227", 4)                = 4
0: read(3, "\vkvs.commit\0}{ \"rootseq\": 282, \""..., 151) = 151
0: write(3, "\0\0\0\264", 4)               = 4
0: write(3, "\0\10cmb.log\0\234{ \"facility\": \"libpmi"..., 180) = 180
0: write(3, "\0\0\0\264", 4)               = 4
0: write(3, "\0\10cmb.log\0\234{ \"facility\": \"libpmi"..., 180) = 180
0: write(3, "\0\0\0Z", 4)                  = 4
0: write(3, "\0\vkvs.commit\0?{ \"ops\": null, \".a"..., 90) = 90
0: read(3, "\0\0\0\227", 4)                = 4
0: read(3, "\vkvs.commit\0}{ \"rootseq\": 283, \""..., 151) = 151
0: write(3, "\0\0\0\264", 4)               = 4
0: write(3, "\0\10cmb.log\0\234{ \"facility\": \"libpmi"..., 180) = 180
0: write(3, "\0\0\0;", 4)                  = 4
0: write(3, "\0\10kvs.get\0#{ \"lwj.8.pmi.TMI_MISM"..., 59) = 59
0: read(3, "\0\0\0:", 4)                   = 4
0: read(3, "\10kvs.get\0#{ \"lwj.8.pmi.TMI_MISMA"..., 58) = 58
```
rank 1 strace
```
1: write(3, "\0\0\0\264", 4)               = 4
1: write(3, "\0\10cmb.log\0\234{ \"facility\": \"libpmi"..., 180) = 180
1: write(3, "\0\0\0Z", 4)                  = 4
1: write(3, "\0\vkvs.commit\0?{ \"ops\": null, \".a"..., 90) = 90
1: read(3, "\0\0\0\227", 4)                = 4
1: read(3, "\vkvs.commit\0}{ \"rootseq\": 282, \""..., 151) = 151
1: write(3, "\0\0\0\264", 4)               = 4
1: write(3, "\0\10cmb.log\0\234{ \"facility\": \"libpmi"..., 180) = 180
1: write(3, "\0\0\0;", 4)                  = 4
1: write(3, "\0\10kvs.get\0#{ \"lwj.8.pmi.TMI_PROV"..., 59) = 59
1: read(3, "\0\0\0B", 4)                   = 4
1: read(3, "\10kvs.get\0+{ \"lwj.8.pmi.TMI_PROVI"..., 66) = 66
1: write(3, "\0\0\0\315", 4)               = 4
1: write(3, "\0\10cmb.log\0\265{ \"facility\": \"libpmi"..., 205) = 205
1: write(3, "\0\0\0Z", 4)                  = 4
1: write(3, "\0\vkvs.commit\0?{ \"ops\": null, \".a"..., 90) = 90
1: read(3, "\0\0\0\227", 4)                = 4
1: read(3, "\vkvs.commit\0}{ \"rootseq\": 283, \""..., 151) = 151
1: write(3, "\0\0\0\264", 4)               = 4
1: write(3, "\0\10cmb.log\0\234{ \"facility\": \"libpmi"..., 180) = 180
1: write(3, "\0\0\0;", 4)                  = 4
1: write(3, "\0\10kvs.get\0#{ \"lwj.8.pmi.TMI_MISM"..., 59) = 59
1: read(3, "\0\0\0:", 4)                   = 4
1: read(3, "\10kvs.get\0#{ \"lwj.8.pmi.TMI_MISMA"..., 58) = 58
```
kvs contents for pmi
```
flux kvs dir lwj.8.pmi
lwj.8.pmi.P1-hostname = opal22
lwj.8.pmi.P0-hostname = opal21
lwj.8.pmi.r2h0 = thnum#2$h0#opal21$r0#0$h1#opal22$r1#1$nbc#1$
lwj.8.pmi.TMI_PROVIDER = psm2(v1.3)
```

**Comment by garlick** [comment]
See also [Intel® Omni-Path Fabric Host Software User Guide](http://www.intel.com/content/dam/support/us/en/documents/network/omni-adptr/sb/Intel_OP_Fabric_Host_Software_UG_H76470_v1_0.pdf)

Runing with `I_MPI_DEBUG=255` gets this:
```
0: [-1] MPI startup(): Imported environment partly inaccesible. Map=0 Info=60cab0
0: [0] MPI startup(): Intel(R) MPI Library, Version 5.1.3  Build 20160120 (build id: 14053)
0: [0] MPI startup(): Copyright (C) 2003-2016 Intel Corporation.  All rights reserved.
0: [0] MPI startup(): Multi-threaded optimized library
1: [-1] MPI startup(): Imported environment partly inaccesible. Map=0 Info=60cab0
0: [0] MPID_nem_impi_create_numa_nodes_map(): Fetching extra numa information from /etc/ofed-mic.map
1: [1] MPID_nem_impi_create_numa_nodes_map(): Fetching extra numa information from /etc/ofed-mic.map
1: [1] MPID_nem_tmi_init(): pg=0x61e740, pg_rank=1
1: [1] init_tmi_library(): TMI lib version is 1.3 
0: [0] MPID_nem_tmi_init(): pg=0x61da00, pg_rank=0
0: [0] init_tmi_library(): TMI lib version is 1.3 
0: [0] init_tmi_provider(): TMI provider initialized: psm2 (1.3)
1: [1] init_tmi_provider(): TMI provider initialized: psm2 (1.3)
```
Launching under slurm (successfully) I get
```
0: [-1] MPI startup(): Imported environment partly inaccesible. Map=0 Info=60cab0
0: [0] MPI startup(): Intel(R) MPI Library, Version 5.1.3  Build 20160120 (build id: 14053)
0: [0] MPI startup(): Copyright (C) 2003-2016 Intel Corporation.  All rights reserved.
0: [0] MPI startup(): Multi-threaded optimized library
1: [-1] MPI startup(): Imported environment partly inaccesible. Map=0 Info=60cab0
0: [0] MPID_nem_impi_create_numa_nodes_map(): Fetching extra numa information from /etc/ofed-mic.map
1: [1] MPID_nem_impi_create_numa_nodes_map(): Fetching extra numa information from /etc/ofed-mic.map
0: [0] MPID_nem_tmi_init(): pg=0x622cc0, pg_rank=0
1: [1] MPID_nem_tmi_init(): pg=0x622a40, pg_rank=1
0: [0] init_tmi_library(): TMI lib version is 1.3 
1: [1] init_tmi_library(): TMI lib version is 1.3 
1: [1] init_tmi_provider(): TMI provider initialized: psm2 (1.3)
0: [0] init_tmi_provider(): TMI provider initialized: psm2 (1.3)
1: [1] init_tmi(): tmi uuid = 0fff0fff-0000-0000-0000-0fff0fff0fff
1: [1] init_tmi(): tmi_ep_open options: TMI_UUID=0fff0fff-0000-0000-0000-0fff0fff0fff, TMI_TNIC_NPEERS=2
0: [0] init_tmi(): tmi uuid = 0fff0fff-0000-0000-0000-0fff0fff0fff
0: [0] init_tmi(): tmi_ep_open options: TMI_UUID=0fff0fff-0000-0000-0000-0fff0fff0fff, TMI_TNIC_NPEERS=2
1: [1] init_tmi(): tmi_ep_open returns 0
1: [1] init_tmi(): tmi_ep_get_addr returns 0
1: [1] init_tmi(): tmi_ep_get_addr: 17-00-00-00-00-00-00-00-01-03-17-00-ff-ff-ff-03, size=16
1: [1] MPID_nem_tmi_vc_init(): vc=0x622b80
1: [1] MPID_nem_tmi_vc_init(): eager_max_msg_size=-1
1: [1] MPID_nem_tmi_init_direct_recv(): direct receive is enabled
0: [0] init_tmi(): tmi_ep_open returns 0
0: [0] init_tmi(): tmi_ep_get_addr returns 0
0: [0] init_tmi(): tmi_ep_get_addr: 76-00-00-00-00-00-00-00-01-03-76-00-ff-ff-ff-03, size=16
0: [0] MPID_nem_tmi_vc_init(): vc=0x622d80
0: [0] MPID_nem_tmi_vc_init(): eager_max_msg_size=-1
0: [0] MPID_nem_tmi_init_direct_recv(): direct receive is enabled
1: [1] MPI startup(): tmi data transfer mode
0: [0] MPI startup(): tmi data transfer mode
1: [1] MPID_nem_tmi_vc_init(): vc=0x622b00
1: [1] MPID_nem_tmi_vc_init(): eager_max_msg_size=-1
0: [0] MPID_nem_tmi_vc_init(): vc=0x622e00
0: [0] MPID_nem_tmi_vc_init(): eager_max_msg_size=-1
1: [1] MPI startup(): Recognition mode: 2, selected platform: 1 own platform: 1
0: [0] MPI startup(): Recognition mode: 2, selected platform: 1 own platform: 1
0: [0] MPI startup(): Device_reset_idx=0
0: [0] MPI startup(): Allgather: 1: 0-4096 & 3-2147483647
0: [0] MPI startup(): Allgather: 1: 16385-2147483647 & 3-4
0: [0] MPI startup(): Allgather: 1: 131073-2147483647 & 17-2147483647
0: [0] MPI startup(): Allgather: 3: 0-2147483647 & 0-2147483647
0: [0] MPI startup(): Allgatherv: 0: 0-2147483647 & 0-2147483647
0: [0] MPI startup(): Allreduce: 1: 0-1024 & 0-16
0: [0] MPI startup(): Allreduce: 1: 0-16384 & 0-4
0: [0] MPI startup(): Allreduce: 3: 32769-262144 & 3-4
0: [0] MPI startup(): Allreduce: 4: 32769-2147483647 & 5-8
0: [0] MPI startup(): Allreduce: 6: 0-2147483647 & 0-2147483647
0: [0] MPI startup(): Alltoall: 1: 0-32 & 17-2147483647
0: [0] MPI startup(): Alltoall: 2: 0-262144 & 3-16
1: [1] MPI startup(): Recognition=2 Platform(code=1 ippn=0 dev=15) Fabric(intra=3 inter=3 flags=0x0)
0: [0] MPI startup(): Alltoall: 2: 524289-2147483647 & 3-4
0: [0] MPI startup(): Alltoall: 2: 33-16384 & 17-2147483647
0: [0] MPI startup(): Alltoall: 4: 262145-2147483647 & 5-16
0: [0] MPI startup(): Alltoall: 3: 0-2147483647 & 0-2147483647
0: [0] MPI startup(): Alltoallv: 1: 0-2147483647 & 0-2147483647
0: [0] MPI startup(): Alltoallw: 0: 0-2147483647 & 0-2147483647
0: [0] MPI startup(): Barrier: 1: 0-2147483647 & 0-2
0: [0] MPI startup(): Barrier: 2: 0-2147483647 & 3-4
0: [0] MPI startup(): Barrier: 3: 0-2147483647 & 5-16
0: [0] MPI startup(): Barrier: 4: 0-2147483647 & 0-2147483647
0: [0] MPI startup(): Bcast: 1: 0-2147483647 & 0-2
0: [0] MPI startup(): Bcast: 1: 0-8192 & 0-89
0: [0] MPI startup(): Bcast: 7: 0-2147483647 & 0-2147483647
0: [0] MPI startup(): Exscan: 0: 0-2147483647 & 0-2147483647
0: [0] MPI startup(): Gather: 1: 0-512 & 0-2
0: [0] MPI startup(): Gather: 2: 65537-262144 & 3-8
0: [0] MPI startup(): Gather: 2: 131073-524288 & 9-32
0: [0] MPI startup(): Gather: 3: 0-2147483647 & 0-2147483647
0: [0] MPI startup(): Gatherv: 1: 0-2147483647 & 0-2147483647
0: [0] MPI startup(): Reduce_scatter: 0: 0-4 & 0-3
0: [0] MPI startup(): Reduce_scatter: 5: 257-512 & 9-16
0: [0] MPI startup(): Reduce_scatter: 1: 0-32768 & 3-2147483647
0: [0] MPI startup(): Reduce_scatter: 1: 262145-524288 & 9-16
0: [0] MPI startup(): Reduce_scatter: 1: 524289-1048576 & 17-32
1: [1] MPID_nem_tmi_iStartContigMsg(): vc 0x622b00, is_dynamic 0, len 48
1: [1] MPID_nem_tmi_get_adi_msg(): Got Adi match: tag 0, rank 0, ctx 32704, size 48, vc 0x622b00 ... receiving
0: [0] MPI startup(): Reduce_scatter: 2: 0-2147483647 & 0-2147483647
0: [0] MPI startup(): Reduce: 2: 129-256 & 0-2
0: [0] MPI startup(): Reduce: 1: 0-2147483647 & 0-2147483647
0: [0] MPI startup(): Scan: 0: 0-2147483647 & 0-2147483647
0: [0] MPI startup(): Scatter: 1: 0-16384 & 0-2
0: [0] MPI startup(): Scatter: 2: 16385-2147483647 & 0-2
0: [0] MPI startup(): Scatter: 2: 1025-2147483647 & 3-2147483647
0: [0] MPI startup(): Scatter: 3: 0-2147483647 & 0-2147483647
0: [0] MPI startup(): Scatterv: 1: 0-2147483647 & 0-2147483647
1: [1] MPID_nem_tmi_iStartContigMsg(): vc 0x622b00, is_dynamic 0, len 48
0: [0] MPI startup(): Rank    Pid      Node name  Pin cpu
0: [0] MPI startup(): 0       70153    opal21     +1
1: [1] MPID_nem_tmi_get_adi_msg(): Got Adi match: tag 0, rank 0, ctx 32704, size 48, vc 0x622b00 ... receiving
1: [1] MPID_nem_tmi_vc_terminate(): vc=0x622b00
0: [0] MPI startup(): 1       55449    opal22     +1
0: [0] MPI startup(): Recognition=2 Platform(code=1 ippn=0 dev=15) Fabric(intra=3 inter=3 flags=0x0)
0: [0] MPI startup(): Topology split mode = 1
0: 
0: | rank | node | space=2
0: |  0  |  0  |
0: |  1  |  1  |
0: [0] MPI startup(): I_MPI_DEBUG=255
0: [0] MPI startup(): I_MPI_FABRICS_LIST=tmi
0: 0: completed MPI_Init in 0.174s.  There are 2 tasks
0: 0: completed first barrier in 0.000s
0: [0] MPID_nem_tmi_iStartContigMsg(): vc 0x622e00, is_dynamic 0, len 48
0: [0] MPID_nem_tmi_get_adi_msg(): Got Adi match: tag 0, rank 1, ctx 32704, size 48, vc 0x622e00 ... receiving
0: [0] MPID_nem_tmi_iStartContigMsg(): vc 0x622e00, is_dynamic 0, len 48
0: [0] MPID_nem_tmi_get_adi_msg(): Got Adi match: tag 0, rank 1, ctx 32704, size 48, vc 0x622e00 ... receiving
0: [0] MPID_nem_tmi_vc_terminate(): vc=0x622e00
1: [1] MPID_nem_tmi_vc_destroy(): vc 0x622b80
0: [0] MPID_nem_tmi_vc_destroy(): vc 0x622d80
1: [1] MPID_nem_tmi_vc_destroy(): vc 0x622b00
0: [0] MPID_nem_tmi_vc_destroy(): vc 0x622e00
0: 0: completed MPI_Finalize in 1.003s
```

**Comment by garlick** [comment]
Still todo before we can close this:  fix the hang when launching Intel PMI jobs under Flux on CTS-1 hardware.  It works on TOSS-3 with the older interconenct (e.g. on upgrade).

**Comment by grondo** [comment]
Following @garlick's instructions above, I'm able to get an Intel MPI run only on a single node to work on opal:
```
(flux-101741) grondo@opal19:~/git/flux-core.git$ FLUX_PMI_TRACE=0xffff flux wreckrun -n2 -N1 t/mpi/hello
0: completed MPI_Init in 0.140s.  There are 2 tasks
0: completed first barrier in 0.000s
0: completed MPI_Finalize in 1.004s
```
2 nodes involved hangs here:
```
(flux-101741) grondo@opal19:~/git/flux-core.git$ FLUX_PMI_TRACE=0xffff flux wreckrun -n2 -N2 t/mpi/hello
Fatal error in MPI_Init: Other MPI error, error stack:
MPIR_Init_thread(784).................: 
MPID_Init(1332).......................: channel initialization failed
MPIDI_CH3_Init(120)...................: 
MPID_nem_init_ckpt(852)...............: 
MPIDI_CH3I_Seg_commit(372)............: 
MPIU_SHMW_Hnd_deserialize(324)........: 
MPIU_SHMW_Seg_open(870)...............: 
MPIU_SHMW_Seg_create_attach_templ(642): open failed - No such file or directory
```

with PMI trace:

```
2016-06-23T17:44:12.829240Z libpmi.debug[0]: PMI_Init = SUCCESS
2016-06-23T17:44:12.829401Z libpmi.debug[0]: PMI_KVS_Get_my_name = SUCCESS
2016-06-23T17:44:12.829842Z libpmi.debug[0]: PMI_KVS_Put (lwj.17.pmi.P0-hostname = "opal19") = SUCCESS
2016-06-23T17:44:12.829862Z libpmi.debug[0]: PMI_KVS_Commit = SUCCESS
2016-06-23T17:44:12.833211Z libpmi.debug[0]: PMI_Barrier (lwj.17.pmi:0, 2) = SUCCESS
2016-06-23T17:44:12.833629Z libpmi.debug[0]: PMI_KVS_Put (lwj.17.pmi.r2h0 = "thnum#1$h0#opal19$r0#0$nbc#1$") = SUCCESS
2016-06-23T17:44:12.833639Z libpmi.debug[0]: PMI_KVS_Commit = SUCCESS
2016-06-23T17:44:12.834803Z libpmi.debug[0]: PMI_Barrier (lwj.17.pmi:1, 2) = SUCCESS
2016-06-23T17:44:12.834823Z libpmi.debug[0]: PMI_Get_rank = SUCCESS
2016-06-23T17:44:12.834830Z libpmi.debug[0]: PMI_Get_size = SUCCESS
2016-06-23T17:44:12.834834Z libpmi.debug[0]: PMI_Get_appnum = SUCCESS
2016-06-23T17:44:12.834839Z libpmi.debug[0]: PMI_KVS_Get_name_length_max = SUCCESS
2016-06-23T17:44:12.834843Z libpmi.debug[0]: PMI_KVS_Get_my_name = SUCCESS
2016-06-23T17:44:12.834859Z libpmi.debug[0]: PMI_KVS_Get_name_length_max = SUCCESS
2016-06-23T17:44:12.834865Z libpmi.debug[0]: PMI_KVS_Get_my_name = SUCCESS
2016-06-23T17:44:12.835855Z libpmi.debug[0]: PMI_KVS_Get_value_length_max = SUCCESS
2016-06-23T17:44:12.835934Z libpmi.debug[0]: PMI_KVS_Get_key_length_max = SUCCESS
2016-06-23T17:44:12.835982Z libpmi.debug[0]: PMI_KVS_Get_value_length_max = SUCCESS
2016-06-23T17:44:12.836013Z libpmi.debug[0]: PMI_KVS_Get_key_length_max = SUCCESS
2016-06-23T17:44:12.836019Z libpmi.debug[0]: PMI_KVS_Get_value_length_max = SUCCESS
2016-06-23T17:44:12.836490Z libpmi.debug[0]: PMI_KVS_Put (lwj.17.pmi.sharedFilename[0] = "/dev/shm/Intel_MPI_PQrIlk") = SUCCESS
2016-06-23T17:44:12.836538Z libpmi.debug[0]: PMI_KVS_Commit = SUCCESS
2016-06-23T17:44:12.837145Z libpmi.debug[0]: PMI_Barrier (lwj.17.pmi:2, 2) = SUCCESS
```

**Comment by grondo** [comment]
Here's a problem, on rank 1 of the 2 node job:
```
2016-06-23T18:11:42.863251Z libpmi.debug[1]: PMI_KVS_Put (lwj.28.pmi.P0-hostname = "opal20") = SUCCESS
```
The second rank of the MPI job seems to think it is rank 0. (Should be using `P1` for rank 1 keys I believe). @garlick do I perhaps have something wrong in my environment? Did 2 node Intel MPI jobs ever work for you?

**Comment by grondo** [comment]
Hmm, this is probably a more generic bug -- I'll open a separate issue
```
$ flux wreckrun -n2 -N2 -l env | grep RANK
1: FLUX_LOCAL_RANKS=1
1: MPIRUN_RANK=0
1: FLUX_TASK_RANK=1
0: FLUX_LOCAL_RANKS=0
0: MPIRUN_RANK=0
0: FLUX_TASK_RANK=0

```

**Comment by grondo** [comment]
Actually perhaps the `MPIRUN_RANK` was a red herring. Intel MPI library doesn't *appear* to look for any `MPIRUN_*` environment variables AFAICS.

However, it *does* seem look for some `SLURM_*` env vars:
```
$ strings /opt/intel-mpi/5.1/intel64/lib/libmpi.so.12 | grep SLURM 
SLURM_PROCID
SLURM_NPROCS
```

I can get a multi-node job to run if I set these variables:

```diff
diff --git a/src/modules/wreck/lua.d/intel_mpi.lua b/src/modules/wreck/lua.d/intel_mpi.lua
index 3447338..88f1c6d 100644
--- a/src/modules/wreck/lua.d/intel_mpi.lua
+++ b/src/modules/wreck/lua.d/intel_mpi.lua
@@ -6,6 +6,12 @@ function rexecd_init ()
     local f = wreck.flux
     local libpmi = f:getattr ('conf.pmi_library_path')
     env['I_MPI_PMI_LIBRARY'] = libpmi
+    env['SLURM_NPROCS'] = wreck.kvsdir.ntasks
 end
 
+function rexecd_task_init ()
+    wreck.environ ['SLURM_PROCID'] = wreck.globalid
+end
+
+
```

```
(flux-101741) grondo@opal19:~/git/flux-core.git$ PMI_DEBUG=0xff I_MPI_DEBUG=7 FLUX_PMI_TRACE=0xff flux wreckrun -n2 -N2 t/mpi/hello
[-1] MPI startup(): Imported environment partly inaccesible. Map=0 Info=60cab0
[0] MPI startup(): Intel(R) MPI Library, Version 5.1.3  Build 20160120 (build id: 14053)
[0] MPI startup(): Copyright (C) 2003-2016 Intel Corporation.  All rights reserved.
[0] MPI startup(): Multi-threaded optimized library
[-1] MPI startup(): Imported environment partly inaccesible. Map=0 Info=60cab0
[0] MPID_nem_tmi_init_direct_recv(): WARNING: I_MPI_TMI_DRECV is not supported in multifabrics mode and will be ignored
[0] MPI startup(): shm and tmi data transfer modes
[1] MPID_nem_tmi_init_direct_recv(): WARNING: I_MPI_TMI_DRECV is not supported in multifabrics mode and will be ignored
[1] MPI startup(): shm and tmi data transfer modes
[1] MPI startup(): Recognition=2 Platform(code=1 ippn=0 dev=13) Fabric(intra=1 inter=3 flags=0x0)
[0] MPI startup(): Device_reset_idx=0
[0] MPI startup(): Allgather: 1: 0-4096 & 3-2147483647
[0] MPI startup(): Allgather: 1: 16385-2147483647 & 3-4
[0] MPI startup(): Allgather: 1: 131073-2147483647 & 17-2147483647
[0] MPI startup(): Allgather: 3: 0-2147483647 & 0-2147483647
[0] MPI startup(): Allgatherv: 0: 0-2147483647 & 0-2147483647
[0] MPI startup(): Allreduce: 1: 0-1024 & 0-16
[0] MPI startup(): Allreduce: 1: 0-16384 & 0-4
[0] MPI startup(): Allreduce: 3: 32769-262144 & 3-4
[0] MPI startup(): Allreduce: 4: 32769-2147483647 & 5-8
[0] MPI startup(): Allreduce: 6: 0-2147483647 & 0-2147483647
[0] MPI startup(): Alltoall: 1: 0-32 & 17-2147483647
[0] MPI startup(): Alltoall: 2: 0-262144 & 3-16
[0] MPI startup(): Alltoall: 2: 524289-2147483647 & 3-4
[0] MPI startup(): Alltoall: 2: 33-16384 & 17-2147483647
[0] MPI startup(): Alltoall: 4: 262145-2147483647 & 5-16
[0] MPI startup(): Alltoall: 3: 0-2147483647 & 0-2147483647
[0] MPI startup(): Alltoallv: 1: 0-2147483647 & 0-2147483647
[0] MPI startup(): Alltoallw: 0: 0-2147483647 & 0-2147483647
[0] MPI startup(): Barrier: 1: 0-2147483647 & 0-2
[0] MPI startup(): Barrier: 2: 0-2147483647 & 3-4
[0] MPI startup(): Barrier: 3: 0-2147483647 & 5-16
[0] MPI startup(): Barrier: 4: 0-2147483647 & 0-2147483647
[0] MPI startup(): Bcast: 1: 0-2147483647 & 0-2
[0] MPI startup(): Bcast: 1: 0-8192 & 0-89
[0] MPI startup(): Bcast: 7: 0-2147483647 & 0-2147483647
[0] MPI startup(): Exscan: 0: 0-2147483647 & 0-2147483647
[0] MPI startup(): Gather: 1: 0-512 & 0-2
[0] MPI startup(): Gather: 2: 65537-262144 & 3-8
[0] MPI startup(): Gather: 2: 131073-524288 & 9-32
[0] MPI startup(): Gather: 3: 0-2147483647 & 0-2147483647
[0] MPI startup(): Gatherv: 1: 0-2147483647 & 0-2147483647
[0] MPI startup(): Reduce_scatter: 0: 0-4 & 0-3
[0] MPI startup(): Reduce_scatter: 5: 257-512 & 9-16
[0] MPI startup(): Reduce_scatter: 1: 0-32768 & 3-2147483647
[0] MPI startup(): Reduce_scatter: 1: 262145-524288 & 9-16
[0] MPI startup(): Reduce_scatter: 1: 524289-1048576 & 17-32
[0] MPI startup(): Reduce_scatter: 2: 0-2147483647 & 0-2147483647
[0] MPI startup(): Reduce: 2: 129-256 & 0-2
[0] MPI startup(): Reduce: 1: 0-2147483647 & 0-2147483647
[0] MPI startup(): Scan: 0: 0-2147483647 & 0-2147483647
[0] MPI startup(): Scatter: 1: 0-16384 & 0-2
[0] MPI startup(): Scatter: 2: 16385-2147483647 & 0-2
[0] MPI startup(): Scatter: 2: 1025-2147483647 & 3-2147483647
[0] MPI startup(): Scatter: 3: 0-2147483647 & 0-2147483647
[0] MPI startup(): Scatterv: 1: 0-2147483647 & 0-2147483647
[0] MPI startup(): Rank    Pid      Node name  Pin cpu
[0] MPI startup(): 0       151593   opal19     +1
[0] MPI startup(): 1       111108   opal20     +1
[0] MPI startup(): Recognition=2 Platform(code=1 ippn=0 dev=13) Fabric(intra=1 inter=3 flags=0x0)
[0] MPI startup(): I_MPI_DEBUG=7
[0] MPI startup(): I_MPI_FABRICS=shm:tmi
[0] MPI startup(): I_MPI_FABRICS_LIST=tmi
0: completed MPI_Init in 0.152s.  There are 2 tasks
0: completed first barrier in 0.000s
0: completed MPI_Finalize in 1.004s
```

**Comment by garlick** [comment]
Well that is progress.

I'd be OK with those changes going into the intel_mpi.lua file for the short term.  Long term I think we'll be able to get Intel MPI to use the PMIv1 wire protocol and then we won't need these (hopefully!)

**Comment by grondo** [comment]
Yeah, I might be missing something obvious but I can't make Intel MPI work multi-node without the above patch to `intel_mpi.lua`. I'll propose this plus MPIRUN_RANK fix in a PR shortly.

**Comment by garlick** [comment]
Bumping this to the 0.5.0 milestone.  It may work at this point, but with with current center power outages we likely can't test anytime soon.

**Comment by garlick** [comment]
Moving back to 0.4.0 milestone as opal is available again.

**Comment by garlick** [comment]
The situation has improved now that wrexecd offers PMI-1 wire protocol service.  I was able to run across multiple nodes if I
* `module load intel-mpi` (in my environment this loads the /opt intel-mpi module)
* unset I_MPI_PMI_LIBRARY (was set by the env module, and points to slurm's libpmi.so)
* comment out SLURM_ environment settings in our `intel_mpi.lua`

Alternatively we can override I_MPI_PMI_LIBRARY to point to ours and that works too.  That's a way to get debug tracing of MPI calls that are in our library, e.g.
```
$ I_MPI_PMI_LIBRARY=/g/g0/garlick/flux-core/build.opal/src/common/.libs/libpmi.so
$ FLUX_PMI_DEBUG=1 flux wreckrun -N2 ./hello.intel
PMI_Init: PMI_FD is set, selecting simple_client
PMI_Init: PMI_FD is set, selecting simple_client
-1: PMI_Init rc=0 
-1: PMI_KVS_Get_my_name ("lwj.5.pmi") rc=0
-1: PMI_Init rc=0 
-1: PMI_KVS_Put ("lwj.5.pmi", "P0-hostname", "opal60") rc=0 
-1: PMI_KVS_Commit rc=0 
-1: PMI_KVS_Get_my_name ("lwj.5.pmi") rc=0
-1: PMI_KVS_Put ("lwj.5.pmi", "P1-hostname", "opal61") rc=0 
-1: PMI_KVS_Commit rc=0 
-1: PMI_Barrier rc=0 
-1: PMI_Barrier rc=0 
-1: PMI_KVS_Get ("lwj.5.pmi", "P1-hostname", "opal61") rc=0
-1: PMI_KVS_Put ("lwj.5.pmi", "r2h0", "thnum#2$h0#opal60$r0#0$h1#opal61$r1#1$nbc#1$") rc=0 
-1: PMI_KVS_Commit rc=0 
-1: PMI_Barrier rc=0 
-1: PMI_Barrier rc=0 
0: PMI_Get_rank rc=0 
-1: PMI_KVS_Get ("lwj.5.pmi", "r2h0", "thnum#2$h0#opal60$r0#0$h1#opal61$r1#1$nbc#1$") rc=0
0: PMI_Get_size rc=0 
1: PMI_Get_rank rc=0 
1: PMI_Get_size rc=0 
0: PMI_Get_appnum rc=0 
0: PMI_KVS_Get_name_length_max rc=0 
0: PMI_KVS_Get_my_name ("lwj.5.pmi") rc=0
0: PMI_KVS_Get_name_length_max rc=0 
0: PMI_KVS_Get_my_name ("lwj.5.pmi") rc=0
0: PMI_KVS_Get_value_length_max rc=0 
0: PMI_KVS_Get_key_length_max rc=0 
0: PMI_KVS_Get_value_length_max rc=0 
1: PMI_Get_appnum rc=0 
1: PMI_KVS_Get_name_length_max rc=0 
1: PMI_KVS_Get_my_name ("lwj.5.pmi") rc=0
1: PMI_KVS_Get_name_length_max rc=0 
0: PMI_Barrier rc=0 
1: PMI_KVS_Get_my_name ("lwj.5.pmi") rc=0
1: PMI_KVS_Get_value_length_max rc=0 
1: PMI_KVS_Get_key_length_max rc=0 
1: PMI_KVS_Get_value_length_max rc=0 
1: PMI_Barrier rc=0 
1: PMI_KVS_Put ("lwj.5.pmi", "P1-businesscard-0", "epaddr_size#16$epaddr#650000000000000001036500FFFFFF03$fabrics_list#shm_and_tmi$") rc=0 
1: PMI_KVS_Commit rc=0 
0: PMI_KVS_Put ("lwj.5.pmi", "P0-businesscard-0", "epaddr_size#16$epaddr#A2000000000000000103A200FFFFFF03$fabrics_list#shm_and_tmi$") rc=0 
0: PMI_KVS_Commit rc=0 
1: PMI_Barrier rc=0 
0: PMI_Barrier rc=0 
1: PMI_KVS_Get_value_length_max rc=0 
0: PMI_KVS_Get_value_length_max rc=0 
1: PMI_KVS_Get ("lwj.5.pmi", "P0-businesscard-0", "epaddr_size#16$epaddr#A2000000000000000103A200FFFFFF03$fabrics_list#shm_and_tmi$") rc=0
0: completed MPI_Init in 0.183s.  There are 2 tasks
0: completed first barrier in 0.000s
1: PMI_Finalize rc=0 
0: PMI_KVS_Get ("lwj.5.pmi", "P1-businesscard-0", "epaddr_size#16$epaddr#650000000000000001036500FFFFFF03$fabrics_list#shm_and_tmi$") rc=0
0: PMI_Finalize rc=0 
0: completed MPI_Finalize in 0.005s
```
Note:
* looks like rank prefix in trace above is -1 when it should be set
* We are still setting FLUX_PMI_LIBRARY_PATH to what is now an invalid path

I'll put together a PR to make it so intel MPI executables just work.

**Comment by grondo** [comment]
I thought we set `I_MPI_PMI_LIBRARY` in environment via the `intel_mpi.lua` wreck plugin. Is this not sufficient due to the other `SLURM` environment variables?

```lua
    local libpmi = f:getattr ('conf.pmi_library_path')
    env['I_MPI_PMI_LIBRARY'] = libpmi
```

**Comment by garlick** [comment]
Oops I think I got confused as I made various changes to the lua module.  Yes, it works with `I_MPI_PMI_LIBRARY` set by the lua module, although it's set to the incorrect path.  Intel PMI then falls back to the wire protocol so it works OK.

IF we fix the path then our libpmi.so is used, which is also a wire protocol client, so functionally the same.

It seems like neither Flux nor slurm (with its new PMI2 default) really need this, and in fact SLURM launch would probably benefit from having it unset.  For now though let's set it to the correct path.
